neoload/commands/report.py

```diff
@@ -0,0 +1,1056 @@
+Quality Status: {{summary.qualityStatus}}
+
+Transactions summary:
+User Path\tElement\tCount\tMin\tAvg\tMax\tPerc 50\tPerc 90\tPerc 95\tPerc 99\tSuccess\tS.Rate\tFailure\tF.Rate
+{% for txn in elements.transactions | rejectattr('id', 'equalto', 'all-transactions') | rejectattr('aggregate.count', 'equalto', '0') | sort(attribute='avgDuration',reverse=true)
+%}{{ txn.user_path|e }}\t{{ txn.name|e }}\t{{ txn.aggregate.count }}\t""" \
+"""{{ txn.aggregate.minDuration }}\t{{ txn.aggregate.avgDuration }}\t{{ txn.aggregate.maxDuration }}\t""" \
+"""{{ txn.aggregate.percentile50 }}\t{{ txn.aggregate.percentile90 }}\t{{ txn.aggregate.percentile95 }}\t{{ txn.aggregate.percentile99 }}\t""" \
+"""{{ txn.aggregate.successCount }}\t{{ txn.aggregate.successRate }}\t""" \
+"""{{ txn.aggregate.failureCount }}\t{{ txn.aggregate.failureRate }}
+{% endfor %}""".strip()
+
+rest_calls = []
+calls_cleanup_indicator = 0
+calls_lock = Lock()
+
+def cleanup_completed_calls():
+    back_step = get_epoch() - (1000)
+    l = list(filter(lambda a: a["completed"] and a["epoch"]<back_step, rest_calls))
+    if len(l) > 20:
+        calls_lock.acquire()
+        logging.getLogger().debug("Cleaning up completed calls")
+        for e in l:
+            try:
+                rest_calls.remove(e)
+            except Exception:
+                logging.getLogger().debug("Clean couldn't remove an e")
+        calls_lock.release()
+
+def get_incomplete_calls():
+    return list(filter(lambda a: a["sent"] and not a["completed"], rest_calls))
+def get_current_calls_per_second_rate():
+    back_step = get_epoch() - (1000)
+    l = len(list(filter(lambda a: a["sent"] and a["epoch"]>=back_step, rest_calls)))
+    return l
+def get_epoch():
+    return time.time() * 1000
+def rest_crud_get(url):
+    global calls_cleanup_indicator
+    global REQUEST_COUNT
+    if calls_cleanup_indicator > 20:
+        cleanup_completed_calls()
+        calls_cleanup_indicator = 0
+
+    REQUEST_COUNT = REQUEST_COUNT + 1
+
+    call = {
+        "uuid": uuid.uuid1(),
+        "url": url,
+        "epoch": get_epoch(),
+        "sent": False,
+        "completed": False,
+    }
+
+    calls_lock.acquire()
+    while math.ceil(get_current_calls_per_second_rate()) >= MAX_CALLS_PER_SECOND:
+        logging.getLogger().debug("Waiting due to rate: rps={}, max={}, total={}".format(
+            str(get_current_calls_per_second_rate()),
+            str(MAX_CALLS_PER_SECOND),
+            REQUEST_COUNT
+        ))
+        time.sleep(0.200)
+
+    rest_calls.append(call)
+    call = next(filter(lambda a: a["uuid"] == call["uuid"],rest_calls))
+
+    logging.getLogger().debug("Rate allows: rps={}, max={}, total={}".format(
+        str(get_current_calls_per_second_rate()),
+        str(MAX_CALLS_PER_SECOND),
+        REQUEST_COUNT
+    ))
+
+    call["sent"] = True
+    calls_cleanup_indicator += 1
+    calls_lock.release()
+    ret = rest_crud.get(url)
+    call["completed"] = True
+    return ret
```

Comment:
What I understand of this code: This is to handle rate limiting for the CLI. It handles rate limit errors and retry every 200ms. But what do we do for calls rejected because of rate limit ? Cancel or postpone ?

- This feature should be in "rest_crud", this is not the responsibility of "report" command to handle api rate limit.

- Did you know that NLW returns a header "Retry-After" and  a json property "retry-after" when limit is reach ? We could use it to just "sleep" until time elapsed.

- The function "get_incomplete_calls" is never used. Remove it.

- Global variable are bad, we should avoid it as much as we can (rest_calls, calls_cleanup_indicator, calls_lock), and they should NEVER be in the middle of the file, always at the beginning.

Reply:
After meeting with Paul, we agree that the CLI "--max-rps" option is a way to avoid making too many requests on NLW. Actuall the limit is 300/min, the CLI set MAX_CALLS_PER_SECOND at 2 by default and manage himself to stay under 2 calls per second.
An improvement could be to handle errors "retry-after" from NLW and retry after the specified time.

Reply:
Agreed for sure, rate-limiting and retry logic should be in rest_crud but didn't want to blow up the team queue. Rather, implementing something local to drive conversation such as retry-after based logic for a later improvement at a global level. I removed unused function, added code comments, and moved global variables to top in c576f96